### PR TITLE
Fixed #32940 -- Removed unnecessary branch and unused squash parameter in Node.add().

### PR DESCRIPTION
**Pull request by kezabelle**
[Ticket 32940](https://code.djangoproject.com/ticket/32940). Let's see if the _whole_ test suite passes or if it will make a liar of me.

### CODE REVIEW

File: django/utils/tree.py [hunk 1]
```diff
@@ -90,11 +90,6 @@ def add(self, data, conn_type, squash=True):
         If `squash` is False the data is prepared and added as a child to
         this tree without further logic.
         """
-        if self.connector == conn_type and data in self.children:
-            return data
-        if not squash:
-            self.children.append(data)
-            return data
         if self.connector == conn_type:
```
Comment (pope1ni) [inline django/utils/tree.py:93]: If we invert this condition we can remove a level of indentation:

```python
    def add(self, data, conn_type):
        """
        Combine this tree and the data represented by data using the connector
        conn_type. The combine is done by squashing the node other away if
        possible.

        This tree (self) will never be pushed to a child node of the combined
        tree, nor will the connector or negated properties change.

        Return a node which can be used in place of data regardless if the node
        other got squashed or not.
        """
        if self.connector != conn_type:
            obj = self._new_instance(self.children, self.connector, self.negated)
            self.connector = conn_type
            self.children = [obj, data]
            return data
        elif (
            isinstance(data, Node) and not data.negated and
            (data.connector == conn_type or len(data) == 1)
        ):
            # We can squash the other node's children directly into this node.
            # We are just doing (AB)(CD) == (ABCD) here, with the addition that
            # if the length of the other node is 1 the connector doesn't
            # matter. However, for the len(self) == 1 case we don't want to do
            # the squashing, as it would alter self.connector.
            self.children.extend(data.children)
            return self
        else:
            # We could use perhaps additional logic here to see if some
            # children could be used for pushdown here.
            self.children.append(data)
            return data
```

**Review by pope1ni — CHANGES_REQUESTED** [review]
@charettes Would you be able to cast an eye over this, particularly because of the removal of the `if self.connector == conn_type and data in self.children:` line which was recently fixed by you to address a longstanding regression. Thanks! 🙂

File: django/utils/tree.py (hunk 1 above)
Comment (kezabelle) [reply django/utils/tree.py:93]: I'll hold on making this change until Simon has had a chance to look (if they so wish), so the diff remains the minimum viable change to evaluate, but yeah the suggestion is probably neater.

Comment (pope1ni) [reply django/utils/tree.py:93]: @kezabelle You could still make this change and it would be functionally identical to the current state of this PR.

The only question is whether we need to restore the following before the `elif` in the code above:

```python
        elif data in self.children:
            return data
```

Note that we don't need to worry about the `self.connector == conn_type` bit that was fixed in b81c7562fc33f50166d5120138d6398dc42b13c3 because, by reordering the code like this, the inverse of that condition can only fall in the first block.

As such, it is clear that this proposed simplification makes the code easier to understand and less prone to the regressions that have gone before.

Comment (charettes) [reply django/utils/tree.py:93]: Thanks for the pings folks.

I'd be very happy to see the `data in self.children` branch go away as it was problematic; it was causing `data.__eq__` to be called for entry in `self.children` which resulted in a huge slowdown during queryset construction.

It's purpose was to prevent queries of the form `Mode.objects.filter(foo='bar').filter(foo='bar')` from resulting in `WHERE foo='bar' AND foo='bar'` but as you can see it's not covered by the suite and has arguable performance benefits since:
1. SQL engines are usually very good at _folding_/optimizing these
2. It's not a very common use case (This branch is reached once by accident in the test suite IIRC)

In other words b81c7562fc33f50166d5120138d6398dc42b13c3 fixed [a flawed optimization with arguable benefits and I'm happy to see it removed](https://code.djangoproject.com/ticket/32632#comment:4).

Comment (kezabelle) [reply django/utils/tree.py]: Appreciate your finding the time to look it over, and I'm pleased we're on the same page.

> SQL engines are usually very good at folding/optimizing these

Yup, they're also far more likely to have the people-bandwidth to actually optimise these kinds of things over time. 

@pope1ni I've replaced the method as per your suggested re-organisation now that Simon has got the gist of the change, and added you to the co-authored yadda yadda for it.

**Comment by felixxm** [comment]
Thanks y'all :1st_place_medal: I split commits.
